### PR TITLE
Allow script to check for new image

### PR DESCRIPTION
**Pull request by grahamhayes**
Only download a new image if the sha does not match

### CODE REVIEW

File: contrib/libvirt/k8s-libvirt.sh [hunk 1]
```diff
@@ -59,15 +61,44 @@ if [ "$1" == "apply" ]; then
     elif [ "$FLAVOUR" == "caasp" ]; then
         IMAGE_PATH="${IMAGE_PATH:-$PWD/SUSE-CaaS-Platform-1.0-KVM-and-Xen.x86_64.qcow2}"
     fi
+    # Get the SHA of the latest file
+    if [ "$FLAVOUR" == "opensuse" ]; then
+        echo "[+] Downloading openSUSE qcow2 VM image sha to '$IMAGE_PATH.sha256.remote'"
+        wget -q -O "$IMAGE_PATH.sha256.remote" -N "http://download.opensuse.org/repositories/Virtualization:/containers:/images:/KVM:/Leap:/42.2/images/Base-openSUSE-Leap-42.2.x86_64-cloud_ext4.qcow2"
```
Comment (aocole) [inline contrib/libvirt/k8s-libvirt.sh]: This looks like it's downloading the qcow2 rather than the sha...

Comment (grahamhayes) [inline contrib/libvirt/k8s-libvirt.sh]: Yup - good catch.

**Comment by grahamhayes** [comment]
@aocole nice catch

**Comment by ereslibre** [comment]
I think `FORCE_IMAGE_REFRESH` and `LATEST_IMAGE` aren't orthogonal, no? Maybe we should rely only on one of the two flags doing the right thing (e.g. I don't want to download the latest image right now, but I get notified there is a new image out there).

**Comment by inercia** [comment]
Do we really need this? Remote images never change: we just generate a new image with a new version. And local images are immutable: they are used as a base for the real images used by libvirt. So I don't see the case where we have to check the _sha_...

**Comment by grahamhayes** [comment]
@inercia yes, we do. if we do not do this, I waste 40 min of my day waiting for a new image that may or may not have changed.

We *should* be versioning them - but we don't - and this tooling (as is) would ignore it if it was.

I am using the signed sha file that is beside the image on the download server, not downloading the full image.

**Comment by inercia** [comment]
> @inercia yes, we do. if we do not do this, I waste 40 min of my day waiting for a new image that may or may not have changed.
> We should be versioning them - but we don't - and this tooling (as is) would ignore it if it was.

AFAIK images do not change in the server, but @thkukuk should know better about this. AFAIK, when a new image is created, the old image is destroyed and a new image with `version = version + 1` is created. And you image cannot change locally in libvirt (unless you are using it in some other way)

> I am using the signed sha file that is beside the image on the download server, not downloading the full image.

Anyway, if you want to just avoid downloading again stuff, I think it would be far better to use something like `if-modified-since` (I've used it with curl, but I'm sure wget supports it too) instead of all this stuff about sha...

**Comment by grahamhayes** [comment]
@inercia I want to be intelligent about when we download. 

There is a generated sha (on the server, pre generated), so when we download the image we download the relevent sha file. 

Then in the future we can just download the 500 byte sha file from the server, to see if the image has changed.

When we download and image, we remove the version from the filename, so we cannot use that.

Using `if-modified-since` would not allow us to add flags to prompt if the image is updated, like @ereslibre suggested.

**Comment by thkukuk** [comment]
@inercia correct, existing images on the server do not change their content. If the content does change, the image gets a new version number.

So there is absolute no need to compare checksums or anything else to define if you need to download something. New number, new image, you should download it. Same number, no need to download it.
You only should check the checksum of the downloaded image, that it is not corrupt.

**Comment by inercia** [comment]
So, as we have said, both files are not going to change. If you want to make sure you avoid downloads for files that do not change, use the `wget` flag that uses `If-Modified-Since`. I don't think all this complexity is necessary, so it is a no from me.

**Comment by thkukuk** [comment]
@grahamhayes if you remove the version number of the image during download, and you later don't know which build this was, how do you want to report a bug if you don't know which image you did test? That's a bad idea.

**Comment by kiall** [comment]
@thkukuk - Agreed, this download tooling is broken - both before and after this PR - it's loosing information by renaming the files and loosing track of what image was downloaded.

With that info lost, we can no longer check "Is this up to date?", and the existing tooling pre-this PR just blindly downloads the latest version no matter what you have locally.. Sucking up developer time waiting for downloads (40 minutes from Ireland, 1hr 40mins for those people in Seattle).

In lieu of totally rewriting things to properly keep track of what image is downloaded, this makes a small change to prevents that time-sink while a better design for actually tracking the content of the downloaded image can be worked out.

**Comment by thkukuk** [comment]
@kiall but then the right fix is to store locally, which version you did download.
How else do you want to report bugs, if you don't know, which build you did test?

**Comment by kiall** [comment]
Thats a good question, and ties into other related areas - like why are always downloading the latest image rather than a known-good image (With a canary CI job checking the latest image, and switching over upon success to ensure we're always as up to date as we can be - without risking a bad salt/terraform etc + image combination blocking developers unnecessarily).

Hence I think it's a bigger thing to discuss and rework, and that a small change to the existing design that saves everyone some time is a good thing.

**Comment by ereslibre** [comment]
While I agree it needs fixing, and the tool is very far from ideal...

> @inercia yes, we do. if we do not do this, I waste 40 min of my day waiting for a new image that may or may not have changed.

> With that info lost, we can no longer check "Is this up to date?", and the existing tooling pre-this PR just blindly downloads the latest version no matter what you have locally.. Sucking up developer time waiting for downloads (40 minutes from Ireland, 1hr 40mins for those people in Seattle).

If you provide `LATEST_IMAGE=false` as [documented here](https://github.com/kubic-project/terraform/tree/master/contrib/libvirt) you won't have this horrible wait times...

**Comment by inercia** [comment]
And if the problem is that we are renaming the image, I think the solution is very simple: fix the renaming.

**Comment by kiall** [comment]
The problem isn't just the renaming, that's 1 reason why the script should be reworked. But, there are others.

Reworking things so that:

* images don't get renamed (or are otherwise somehow tracked)
* don't get unnecessarily downloaded when you already have the exact file locally
* don't get downloaded unless a CI run has passed with it (See the dockershim issue from a few days ago)
* other parts of this repo to not assume a static filename
* ensuring old images are purged from libvirt, to avoid filling everyones harddisk

Is something that should be discussed, and planned. In the meantime, this small change to 1 file prevents the timesink of re-downloding images.

@ereslibre yes, LATEST_IMAGE=false will prevent the download. I run with that every single time. However, once I hit a failure.. I now have no idea if it's because my image is out of date or not.. So I have to either redeploy with the latest image or spend the time to investigate if a new image is necessary.

This change means I can always run with LATEST_IMAGE=true, and not have to think about updating images..

I don't think anyone is arguing the download code is perfect before or after this PR, or that this PR doesn't improve the situation.

I'm not sure I understand the issue?

**Comment by grahamhayes** [comment]
Added the extra prompt flags as requested by @ereslibre

**Comment by grahamhayes** [comment]
This was a small fix to save developers who are not in an office with a local mirror a lot of time.

I agree, we should be doing things very differently when dealing with images, but that is a bigger change (e.g. We should be pinning to a version in this repo, so when we change to a new version of the image we actually run CI on it).

All of these changes are larger, and require editing terraform templates.

If there is massive opposition to this time saving measure for remote developers, I can carry it as a local patch, and send patch files around to people who need it.

File: contrib/libvirt/k8s-libvirt.sh [hunk 2]
```diff
@@ -60,14 +73,62 @@ if [ "$1" == "apply" ]; then
...
+
+        # Compare the current local SHA to the remote one
+        cmp --silent $LOCAL_SHA $REMOTE_SHA && NEED_UPDATE=false || NEED_UPDATE=true
+
+    fi
+
+    if $PROMPT_LATEST_IMAGE && $NEED_UPDATE; then
+
+        if which notify-send >/dev/null; then
+            notify-send "There is a new $FLAVOUR image availible - should we update?"
```
Comment (ereslibre) [inline contrib/libvirt/k8s-libvirt.sh]: Typo in `availible`. All the rest too :)

File: contrib/libvirt/k8s-libvirt.sh [hunk 3]
```diff
@@ -92,7 +153,7 @@ fi
 [ -n "$TF_DEBUG" ] && export TF_LOG=debug
 [ -n "$FORCE" ] && force="--force"
 
-# Go kubes go!
+Go kubes go!
```
Comment (ereslibre) [inline contrib/libvirt/k8s-libvirt.sh]: Oops...

File: contrib/libvirt/k8s-libvirt.sh (hunk 2 above)
Comment (grahamhayes) [inline contrib/libvirt/k8s-libvirt.sh]: Fixed :)

**Comment by inercia** [comment]
This script has always been a massive mess and I think this makes it even worse. But I must admit that this PR sticks to the style previously found in the code: not a single function, no constants whatsoever, etc... ;)

**Comment by kiall** [comment]
@inercia I agree, bash is already a "bad" enough language to code in, that care should be taken that the code written is understandable and clean... However, blocking a change that is actively wasting developer time, and which forms no part of the shipped product, on a redesign / refactor does not make sense to me.

Based on the history of this script, it evolved from being 1 persons personal handy script to THE way to develop CaaSP, it appears to have served its purpose well, but has enough shortcomings that a rethink is likely due. That should be designed and planned, not simply included as part of a minor fix to the repo.

**Comment by grahamhayes** [comment]
Sure - that is what a good developer does - when editing code in a pre existing file, if you are not completely refactoring / replacing it, you stick to the same style.

This script doesn't even for for 75% of the use cases it is supposed to, but provides useful information and saves time for people at the end of a VPN connection.

Going forward I can start a conversation about how to replace this script so it works, and we can hash out how we do that as an actual design.

**Comment by PI-Victor** [comment]
ysd

**Comment by flavio** [comment]
Ok, this is a really really long thread for such a small change.

I think we all agree about:

  * the current script should be revisited to be written in a less hackish way
  * the script should do things in a different way ([this](https://github.com/kubic-project/terraform/pull/57#issuecomment-299160214) comment seems to provide a nice overview)
  * people are wasting a lot of time downloading stuff

I'm going to merge this PR because it addresses the immediate issue (the download time) and because a proper fix requires more time and a **reasonable** amount of discussion to design a better solution.

Can someone create a Trello card to address the issue and can the people involved in this discussion join it when it will be discussed?

**Pull request merged**